module/block_iterator/creator_test.go

```diff
@@ -0,0 +1,334 @@
+package block_iterator
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+)
+
+// TestCanIterateNewLatest: iterate through all heights from root to latest,
+// 			and after finish iteration, iterate again with an updated latest height,
+// 			verify it can iterate from latest + 1 to new latest
+// TestCanResume: stop at a height, and take checkpoint, create a new iterator,
+// 			verify it will resume from the next height to the latest
+// TestCanSkipViewsIfNotIndexed: iterate through all views, and skip views that are not indexed
+
+func TestCanIterate(t *testing.T) {
+	root := &flow.Header{Height: 0}
+	// Create mock blocks
+	blocks := []*flow.Header{
+		{Height: 1},
+		{Height: 2},
+		{Height: 3},
+		{Height: 4},
+		{Height: 5},
+	}
+
+	// Mock getBlockIDByHeight function
+	getBlockIDByHeight := func(height uint64) (flow.Identifier, error) {
+		for _, block := range blocks {
+			if block.Height == height {
+				return block.ID(), nil
+			}
+		}
+		return flow.Identifier{}, fmt.Errorf("block not found at height %d", height)
+	}
+
+	// Mock progress tracker
+	progress := &mockProgress{}
+
+	// Mock latest functions
+	latest := func() (*flow.Header, error) {
+		return blocks[len(blocks)-1], nil
+	}
+
+	// Create iterator
+	creator, err := NewHeightBasedCreator(
+		getBlockIDByHeight,
+		progress,
+		root,
+		latest,
+	)
+	require.NoError(t, err)
+
+	iterator, err := creator.Create()
+	require.NoError(t, err)
+
+	// Iterate through blocks
+	visitedBlocks := make([]flow.Identifier, 0, len(blocks))
+	for {
+		blockID, ok, err := iterator.Next()
+		require.NoError(t, err)
+		if !ok {
+			break
+		}
+
+		visitedBlocks = append(visitedBlocks, blockID)
+	}
+
+	// Verify all blocks were visited exactly once
+	for i, block := range blocks {
+		require.Equal(t, block.ID(), visitedBlocks[i], "Block %v was not visited", block.Height)
+	}
+
+	// Verify no extra blocks were visited
+	require.Equal(t, len(blocks), len(visitedBlocks), "Unexpected number of blocks visited")
+
+	// Verify the final checkpoint
+	next, err := iterator.Checkpoint()
+	require.NoError(t, err)
+	require.Equal(t, uint64(6), next, "Expected next height to be 6 (last height + 1)")
+	savedNextHeight, err := progress.ProcessedIndex()
+	require.NoError(t, err)
+	require.Equal(t, uint64(6), savedNextHeight, "Expected next height to be 6 (last height + 1)")
+
+	// Additional blocks to be added later
+	additionalBlocks := []*flow.Header{
+		{Height: 6},
+		{Height: 7},
+		{Height: 8},
+	}
+	visitedBlocks = make([]flow.Identifier, 0, len(additionalBlocks))
+
+	// Update blocks so that the latest block is updated, and getBlockIDByHeight
+	// will return the new blocks
+	blocks = append(blocks, additionalBlocks...)
+
+	// Create another iterator
+	iterator, err = creator.Create()
```

Comment:
Creator is reused to create new iteration after previous iteration is completed. So when chunk data pack pruner, after pruning to latest block, the creator can create a new iterator to prune to the new latest height.